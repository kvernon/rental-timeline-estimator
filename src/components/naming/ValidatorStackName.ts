import { formatName } from './FormatName';
import { FormatNames } from './FormatNames';

export const ValidatorStackName = (id: string) => formatName(id, FormatNames.ValidatorStackId);
