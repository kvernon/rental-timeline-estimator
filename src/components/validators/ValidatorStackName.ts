export const ValidatorStackName = (id: string) => `validatorStack${id}`;
