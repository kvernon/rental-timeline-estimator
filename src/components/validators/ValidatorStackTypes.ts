export enum ValidatorStackTypes {
  Optional,
  Required,
}
