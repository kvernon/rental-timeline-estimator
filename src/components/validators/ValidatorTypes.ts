export enum ValidatorTypes {
  Invalid = 0,
  Valid = 1,
  Optional = 2,
}
