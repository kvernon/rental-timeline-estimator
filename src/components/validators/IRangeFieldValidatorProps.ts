import { ValidatorStackTypes } from './ValidatorStackTypes';
import { IRangeFieldValidatorBlur } from './IRangeFieldValidatorBlur';
import { FontGroups } from '../../theming/fontGroups';

export interface IRangeFieldValidatorProps extends IRangeFieldValidatorBlur {
  inputFontGroup?: FontGroups;
  inputLabelFontGroup?: FontGroups;
  validationType: ValidatorStackTypes;
  direction?: 'row' | 'column';
  title?: string;
  prefix?: string;
  suffix?: string;
  /**
   * will default to 0
   */
  min?: number;

  /**
   * will default to 100
   */
  max?: number;

  /**
   * will default to 0
   */
  defaultValue?: number;

  hasSpinner?: boolean;

  useUnderlineOnly?: boolean;

  useTransparent?: boolean;
}
