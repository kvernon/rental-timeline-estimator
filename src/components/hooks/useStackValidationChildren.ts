import { ValidatorStackTypes } from '../validators/ValidatorStackTypes';
import React, { useMemo, useState } from 'react';
import { ValidatorTypes } from '../validators/ValidatorTypes';
import { useChildrenValueNames } from './useChildrenValueNames';
import { useChildrenIdsList } from './useChildrenIdsList';
import { useValid } from './useValid';
import { useWatcher } from './useWatcher';
import { IIdentifier } from '../validators/IIdentifier';

const translateValues = (names: string[], valid: ValidatorTypes) => {
  return names.map((c) => {
    return {
      name: c,
      result: valid,
    };
  });
};

export const useStackValidationChildren = (
  panelValidatorStackType: ValidatorStackTypes = ValidatorStackTypes.Optional,
  children: React.ReactElement<IIdentifier>[] | React.ReactElement<IIdentifier> = [],
) => {
  const [theChildrenIds] = useChildrenIdsList(children);

  const [isValid] = useValid(panelValidatorStackType);

  const [childrenValidationResultNames] = useChildrenValueNames(theChildrenIds, 'validationResult');

  const [isValidCollection, setIsCollectionValid] = useState<ValidatorTypes[]>([]);

  useMemo(() => {
    if (childrenValidationResultNames.length !== isValidCollection.length) {
      setIsCollectionValid(Array(childrenValidationResultNames.length).fill(isValid));
    }
  }, [childrenValidationResultNames, isValid, isValidCollection]);

  const [isAllValid, setIsAllValid] = useState<ValidatorTypes>(isValid);

  const [watcherResults] = useWatcher<ValidatorTypes>(childrenValidationResultNames);

  useMemo(() => {
    if (watcherResults.some((x) => x !== undefined)) {
      const newOne = [...isValidCollection];
      watcherResults.forEach((value: ValidatorTypes, idx: number) => {
        if (value !== undefined) {
          newOne[idx] = value;
        }
      });

      if (isValidCollection.join(',') !== newOne.join(',')) {
        setIsCollectionValid(newOne);
      }
    }
  }, [watcherResults, isValidCollection]);

  const [finalResult, setFinalResult] = useState<
    {
      name: string;
      result: ValidatorTypes;
    }[]
  >([]);

  useMemo(() => {
    setIsAllValid(() => {
      if (isValidCollection.some((x) => x === ValidatorTypes.Invalid)) {
        return ValidatorTypes.Invalid;
      }

      if (isValidCollection.some((x) => x === ValidatorTypes.Valid)) {
        return ValidatorTypes.Valid;
      }

      return ValidatorTypes.Optional;
    });

    const r = translateValues(childrenValidationResultNames, isValid);
    isValidCollection.forEach((x, index) => {
      r[index].result = x;
    });

    if (JSON.stringify(r) !== JSON.stringify(finalResult)) {
      setFinalResult(r);
    }
  }, [finalResult, childrenValidationResultNames, isValidCollection, isValid]);

  return {
    isValidCollection: finalResult,
    isValid: isAllValid,
  };
};
