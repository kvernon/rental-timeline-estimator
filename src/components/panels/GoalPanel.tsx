import React, { useEffect, useState } from 'react';

import { RangeFieldValidator } from '../validators/RangeFieldValidator';
import { IRangeFieldValidatorProps } from '../validators/IRangeFieldValidatorProps';
import styled from '@emotion/styled';
import { Spinner } from '../core/Spinner';
import { Stack } from '../core/Stack';

const TopStack = styled.div`
  height: 275px;
  display: block;
  width: 100%;
  overflow: visible;
  float: left;
  position: relative;
`;

const StackPosition = styled(Stack)`
  position: relative;
  width: auto;
  min-height: 0;
  z-index: 2;
`;

export const GoalPanel = <Required extends boolean = false>(props: IRangeFieldValidatorProps<Required>) => {
  const [x, setX] = useState<number | undefined>();
  const [y, setY] = useState<number | undefined>();
  const [width, setWidth] = useState<number | undefined>();
  const [height, setHeight] = useState<number | undefined>();

  const ZRangeFieldValidator = styled(RangeFieldValidator<Required>)`
    z-index: 4;
  `;

  useEffect(() => {
    const getPosition = () => {
      const ele = document.getElementById(`${props.id}-box`);

      const x = ele?.getBoundingClientRect().x;
      setX(() => x);

      const y = ele?.getBoundingClientRect().y;
      setY(() => y);

      const width = ele?.getBoundingClientRect()?.width;
      setWidth(() => width);

      const height = ele?.getBoundingClientRect()?.height;
      setHeight(() => height);
    };

    getPosition();
    window.addEventListener('resize', getPosition);

    return () => {
      window.removeEventListener('resize', getPosition);
    };
  }, [props.id]);

  return (
    <TopStack aria-label={'Goal Panel'}>
      <Spinner shape={{ x, width, y, height }} />
      <StackPosition spacing={2} paddingLeft={'20%'} paddingTop={'25px'} paddingBottom={'25px'} paddingRight={'20%'}>
        <ZRangeFieldValidator
          {...props}
          id={props.id}
          direction={'row'}
          useUnderlineOnly
          hasSpinner={false}
          useTransparent={true}
          required={props.required}
        />
      </StackPosition>
    </TopStack>
  );
};
