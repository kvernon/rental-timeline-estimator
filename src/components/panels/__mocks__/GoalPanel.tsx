import React from 'react';

import { IRangeFieldValidatorProps } from '../../validators/IRangeFieldValidatorProps';
import { ValidatorTypes } from '../../validators/ValidatorTypes';

export const GoalPanel = jest.fn((p: IRangeFieldValidatorProps<true>) => {
  return (
    <input
      type="number"
      id={p.id}
      value={p.value?.value}
      name={p.title}
      aria-label={'Goal Panel'}
      onChange={(e) => {
        if (p.onChange)
          p.onChange({
            value: parseInt(e.target.value),
            validationResult: ValidatorTypes.Valid,
          });
      }}
    />
  );
});
