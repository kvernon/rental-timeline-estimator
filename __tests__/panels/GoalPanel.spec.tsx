import { configure, render, screen } from '@testing-library/react';
import { GoalPanel } from '../../src/components/panels/GoalPanel';
import React from 'react';
import { IRangeFieldValidatorProps } from '../../src/components/validators/IRangeFieldValidatorProps';
import { ValidatorStackTypes } from '../../src/components/validators/ValidatorStackTypes';
import { useStackValidationChildren } from '../../src/components/hooks/useStackValidationChildren';
import { ValidatorTypes } from '../../src/components/validators/ValidatorTypes';
import { ValidatorStackName } from '../../src/components/naming/ValidatorStackName';

jest.mock('../../src/components/validators/RangeFieldValidator', () => {
  return {
    RangeFieldValidator: jest.fn((p: IRangeFieldValidatorProps) => <input type="number" id={p.id} value={p.defaultValue} name={p.title}></input>),
  };
});
jest.mock('react-hook-form', () => {
  //const watchMock: jest.Mock = jest.fn();
  return {
    useForm: jest.fn().mockReturnValue({}),
    useFormContext: jest.fn().mockReturnValue({
      register: jest.fn(),
      getValues: jest.fn(),
      setValue: jest.fn(),
      trigger: jest.fn(),
    }),
  };
});

jest.mock('../../src/components/hooks/useStackValidationChildren');

describe('GoalPanel unit test', () => {
  afterEach(() => {
    jest.resetModules();
  });

  beforeEach(() => {
    configure({ testIdAttribute: 'id' });
  });

  test('should render', () => {
    const useValidationChildrenMocked = jest.mocked(useStackValidationChildren);
    useValidationChildrenMocked.mockReturnValueOnce({
      isValid: ValidatorTypes.Optional,
      isValidCollection: [],
    });

    const p: IRangeFieldValidatorProps = {
      id: 'goal',
      validationType: ValidatorStackTypes.Required,
    };

    render(<GoalPanel {...p} />);

    const entity = screen.getByTestId<HTMLDivElement>(ValidatorStackName(p.id));

    expect(entity).toMatchSnapshot();
  });
});
