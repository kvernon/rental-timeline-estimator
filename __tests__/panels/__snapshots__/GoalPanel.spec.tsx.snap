// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalPanel unit test should render 1`] = `
.emotion-0 {
  height: 275px;
  display: block;
  width: 100%;
  overflow: visible;
  float: left;
  position: relative;
}

.emotion-1 {
  width: 950px;
  height: 950px;
  left: 240px;
  top: -332px;
  float: left;
  position: absolute;
  background-image: radial-gradient(rgba(199, 159, 0, 0.3) 0%, rgba(61, 61, 61, 0.01) 62%, transparent 67%);
  border-radius: 50%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  z-index: -1;
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 300s;
  animation-duration: 300s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
}

.emotion-1::after,
.emotion-1::before {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  transform-origin: center;
  border-radius: 50%;
  -webkit-mask-image: radial-gradient(rgba(0, 0, 0, 1) 65%, transparent 80%);
  mask-image: radial-gradient(rgba(0, 0, 0, 1) 65%, transparent 80%);
}

.emotion-1::before {
  background: repeating-conic-gradient(from 0deg, rgb(199, 159, 0, 0.3) 0deg 10deg, transparent 10deg 15deg);
  -webkit-mask-image: radial-gradient(rgba(0, 0, 0, 0.7) 0%, rgba(0, 0, 0, 0.3) 36%, transparent 65%);
  mask-image: radial-gradient(rgba(0, 0, 0, 0.7) 0%, rgba(0, 0, 0, 0.3) 36%, transparent 65%);
  -webkit-animation: rotate 720s linear,scale 3s linear infinite;
  animation: rotate 720s linear,scale 3s linear infinite;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
  padding-bottom: 25px;
  margin-bottom: 0;
  padding-top: 25px;
  padding-left: 20%;
  padding-right: 20%;
  position: relative;
  width: auto;
  min-height: 0;
  z-index: 2;
}

<div
  class="emotion-0"
  id="validatorStackgoal"
>
  <div
    class="emotion-1"
    shape="[object Object]"
  />
  <div
    class="emotion-2"
    id="goal"
    spacing="2"
  >
    <input
      id="goal"
      type="number"
      value=""
    />
  </div>
</div>
`;
